### PR TITLE
Fix bug

### PR DESCRIPTION
**Pull request by rakkit**

### CODE REVIEW

**Review by rakkit — COMMENTED** [review]
you can ignore this commit [9ddd725f25e5b9e776769a735d9165f0a2d55eeb]

File: torchtitan/models/MoEllama/moellama.py [hunk 1]
```diff
@@ -366,6 +373,57 @@ def init_weights(
         init_fn(self.down_proj.weight, mean=0.0, std=gate_init_std)
 
 
+@torch.compile(dynamic=True)
+def mixed_experts_output(
```
Comment (rakkit) [inline torchtitan/models/MoEllama/moellama.py]: This will be dropped in a later commit.
same as `prefill_buffer` and `execute_permute`

MoE's forward will be rewritten

File: torchtitan/models/MoEllama/moellama.py [hunk 2]
```diff
@@ -785,9 +930,16 @@ def forward(
             torch.Tensor: Output tensor after applying attention and feedforward layers.
 
         """
-        h = x + self.attention(self.attention_norm(x), freqs_cis)
 
-        mlp_output, moe_aux_loss, routing_entropy = self.feed_forward(self.ffn_norm(h))
+        @torch.compile(fullgraph=True)
```
Comment (rakkit) [inline torchtitan/models/MoEllama/moellama.py:838]: this is helpful to enhance MoE's compile

**Review by rakkit — COMMENTED** [review]
this is important

File: torchtitan/components/optimizer.py
Comment (rakkit) [inline torchtitan/components/optimizer.py:1]: commit [336e10be10bc8466fa636953bfc14c8072b4c900]

1. MoE's weights should be automatically sent to "experts", the device mesh does not matter.
2. MoE's weights is [G, D-in, D-out], but optimzier/norm logic is [D-out, D-in].  need flag to tranpose the weight/gradient

3. To speed up Norm log for MoE. we only run norm_func on local parameters and Gradient. and we do all-gather to log on rank-0

File: torchtitan/models/MoEllama/__init__.py
Comment (rakkit) [inline torchtitan/models/MoEllama/__init__.py:1]: after fix bugs on moe's init and update, we can have `norm-every-where` now, and set it to `True` by default to aglin dense model

File: torchtitan/models/MoEllama/moellama.py [hunk 3]
```diff
@@ -871,6 +766,9 @@ def __init__(self, layer_id: int, model_args: MoEModelArgs):
                 )
         else:
             routed_scaling_factor = model_args.moe_routed_scaling_factor
+            logger.info(
+                f"Using manually set routed_scaling_factor: {routed_scaling_factor}"
+            )
```
Comment (rakkit) [inline torchtitan/models/MoEllama/moellama.py]: the `buffer` `moe_routed_scaling_factor` will be filled by 0 in meta_device init.  need to manually fill up

File: torchtitan/components/optimizer.py [hunk 4]
```diff
@@ -382,6 +385,7 @@ def get_parameter_norms(self):
                         "eps": group["eps"],
                         "norm_factor": group["norm_factor"],
                         "zeropower_backend": group["backend"],
+                        "zeropower_backend": group["backend"],
```
Comment (rakkit) [inline torchtitan/components/optimizer.py]: idk why i repeat it twice

File: torchtitan/models/MoEllama/moellama.py [hunk 5]
```diff
@@ -24,12 +24,11 @@
 from torchtitan.experiments.kernels.moe.indices import generate_permute_indices
 
 from .GroupedExperts import GroupedExperts
+from grouped_gemm.ops import permute, unpermute
```
Comment (rakkit) [inline torchtitan/models/MoEllama/moellama.py:27]: maybe better to add a try-catch here
```
try:
    from grouped_gemm.ops import permute, unpermute

```
incase someone want to test othercode (eg dense llama) but forget to install kernels

File: torchtitan/optimizers/distributed_scion.py [hunk 6]
```diff
@@ -47,9 +47,12 @@ def get_param_type(p, fsdp_enabled, expert_enabled):
 
 def calculate_shard_shape(shape, rank, world_size):
     full = shape[0]
-    base = full // world_size
-    extra = full % world_size
-    dim0 = base + 1 if rank < extra else base
+    splits = torch.arange(full).chunk(world_size)
```
Comment (rakkit) [inline torchtitan/optimizers/distributed_scion.py:50]: The original calculation is buggy.  FSDP document says it follows the logic of torch.chunk(dim=0).

example:  shape=5, world_size=4,
original (wrong) gives [2, 1, 1, 1]
FSDP2 and now (correct) gives [2, 2, 1, 0]

File: torchtitan/optimizers/distributed_scion.py [hunk 7]
```diff
@@ -222,13 +225,19 @@ def _lmo_for_2d_tensor(g, need_transpose=False):
...
             g = _lmo_for_2d_tensor(g, need_transpose=is_grouped_experts)
         elif g.ndim == 3:
-            g = torch.stack(
-                [
-                    _lmo_for_2d_tensor(g[i], need_transpose=is_grouped_experts)
-                    for i in range(g.shape[0])
-                ],
-                dim=0,
-            )
+            if g.shape[0] > 0:
```
Comment (rakkit) [inline torchtitan/optimizers/distributed_scion.py:228]: when g.shape[0] is 0, do  `g = torch.stack([ ])` will throw error

File: torchtitan/components/optimizer.py [hunk 8]
```diff
@@ -326,11 +330,13 @@ def compute_grad(p, optimizer=None, **kwargs):
                 momentum,
                 nesterov,
                 update_buffer=False,
-                gather_to_local=optimizer.fsdp_enabled,
+                gather_to_local=optimizer.fsdp_enabled and p.ndim < 3,
```
Comment (rakkit) [inline torchtitan/components/optimizer.py:333]: we dont gather MOE's weights here, otherwise all rank will get full of [G, d-in, d-out] for `p`, but [G/FSDP, d-in, d-out] for `g`

**Comment by janEbert** [comment]
Merged.
